### ISSUE TITLE
F-Droid reproducible build failed

### ISSUE THREAD

**Issue by linsui** [pewaru-333/HomeMedkit-App#79]
Was your apk built from the tag? Maybe you need to clean the cache and rebuild it.

**Comment by pewaru-333** [comment]
@linsui, hello. Yes, I did it like always. Do I need to attempt again for this update?

**Comment by linsui** [comment]
[ru.application.homemedkit_46.apk.html.zip](https://github.com/user-attachments/files/18250296/ru.application.homemedkit_46.apk.html.zip)

Do you have any idea about the diff? Maybe you can try to rebuild it. Are there any toolchain changed?

**Comment by pewaru-333** [comment]
@linsui, the only thing, that come to my mind, is that I added automatic generation of locales file (https://developer.android.com/guide/topics/resources/app-languages#auto-localeconfig).

I can rebuild .apk now, do I need to change the file in last release tag?

**Comment by linsui** [comment]
Yes, please change the file in the last release if the apk is different. Thanks!

**Comment by pewaru-333** [comment]
@linsui, I have rebuild .apk and pinned it to the last release.

**Comment by linsui** [comment]
Thanks!

**Comment by linsui** [comment]
Can you clean the cache and rebuild 1.8.2 apk? Thanks!

**Comment by pewaru-333** [comment]
@linsui, hi! Have done it 1 minute ago.

**Comment by linsui** [comment]
It's still not reproducible. 

[ru.application.homemedkit_57.apk.html.zip](https://github.com/user-attachments/files/21328877/ru.application.homemedkit_57.apk.html.zip)

Could you please take a look?

**Comment by pewaru-333** [comment]
I don't know, why has it happened, because I edited just 10 lines of code in another place... I will make a very small update next week to overcome this issue!

**Comment by linsui** [comment]
It's not fixed. Those DayOfWeek need to be sorted. I guess there is somewhere a list.

**Comment by pewaru-333** [comment]
@linsui, it is Java enum. I use it when adding to database and to update UI state. But take a look at screenshot: on the left side, where the build is generated, days and their ordinal are in chaotic order for some reason. Can it be a bug?

I have just builded the app, I got this code in build folder in Kotlin:

```kotlin
private fun __DayOfWeek_enumToString(_value: DayOfWeek): String = when (_value) {
    DayOfWeek.MONDAY -> "MONDAY"
    DayOfWeek.TUESDAY -> "TUESDAY"
    DayOfWeek.WEDNESDAY -> "WEDNESDAY"
    DayOfWeek.THURSDAY -> "THURSDAY"
    DayOfWeek.FRIDAY -> "FRIDAY"
    DayOfWeek.SATURDAY -> "SATURDAY"
    DayOfWeek.SUNDAY -> "SUNDAY"
  }
```

which is then complied into Java bytecode with JVM somehow, and the code on the right side of [screenshot](https://private-user-images.githubusercontent.com/147328717/469381988-94d49347-a78f-44f4-b819-8b12411c2552.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMyMTI2NTMsIm5iZiI6MTc1MzIxMjM1MywicGF0aCI6Ii8xNDczMjg3MTcvNDY5MzgxOTg4LTk0ZDQ5MzQ3LWE3OGYtNDRmNC1iODE5LThiMTI0MTFjMjU1Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcyMlQxOTI1NTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wMjAzMGZlYmU5OWI0ZWFlNzNlYTk4YTUxY2RlYThjZTZlZDZkZjBmNWVkNDQ3ODM4NTllZDk4YjdhMDBkN2EyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.epIfjyzDQxtpXpdclriEEug11KT23u5PbRRrh5Z8stM) is the Java "analogue" of Kotlin code in the block upper here.

Can you find a share what was in app versio 1.8.1 at this place?

**Comment by linsui** [comment]
So the generated file is still in an order but the bytecode is not?

> Can you find a share what was in app versio 1.8.1 at this place?

Do you mean this file in our build? Why 1.8.1?

**Comment by pewaru-333** [comment]
> So the generated file is still in an order but the bytecode is not?

Yes, according to that report that you have attached.

> Do you mean this file in our build? Why 1.8.1?

In your build. Because it was last RB.

I not in a build .apk a lot, just know how to hit buttons and attach here. :smiley::smiley::smiley:

**Comment by pewaru-333** [comment]
> I tried to rebuild 1.8.1 but it's not reproducible.

This is strange.

> Maybe related to [this one](https://github.com/pewaru-333/HomeMedkit-App/blob/master/app/src/main/java/ru/application/homemedkit/data/MedicineDatabase.kt#L526)?

Probably not, because I was added on March 28, update was RB. Sorting is not affected here, because it is built in construction with order, like "Monday - 1, Tuesday - 2 ...".

I have decomplied attached .apk, what I have:

```java
public final String __DayOfWeek_enumToString(DayOfWeek dayOfWeek) {
        switch (WhenMappings.$EnumSwitchMapping$0[dayOfWeek.ordinal()]) {
            case 1:
                return "MONDAY";
            case 2:
                return "TUESDAY";
            case 3:
                return "WEDNESDAY";
            case 4:
                return "THURSDAY";
            case 5:
                return "FRIDAY";
            case 6:
                return "SATURDAY";
            case 7:
                return "SUNDAY";
            default:
                throw new NoWhenBranchMatchedException();
        }
    }
```

Can it be related to Gradle version? [Here](https://github.com/pewaru-333/HomeMedkit-App/blob/master/gradle/libs.versions.toml#8) is the Gradle plugin version, and [here](https://github.com/pewaru-333/HomeMedkit-App/blob/master/gradle/wrapper/gradle-wrapper.properties) Gradle's version itself.

**Comment by pewaru-333** [comment]
@IzzySoft, hi! I think it was Gradle version issue, Also, I have setup a workflow to build .apk directly from repository code. Not expected to face this issue again!

**Comment by IzzySoft** [comment]
@linsui 1.8.5 was RB here. Currently running 1.8.6 to check… RB confirmed for this, too. 🤷‍♂️ So what exactly does happen on your end?

```yaml
- ./gradlew assembleRelease
- git clone -b v0.3.0 https://github.com/obfusk/reproducible-apk-tools.git
- reproducible-apk-tools/zipalign.py --page-size 16 --pad-like-apksigner --replace app/build/outputs/apk/release/*unsigned.apk /outputs/unsigned.apk
```

I miss the adjustment for apksigner padding in your recipe. Needed since v1.8.4. Could it be that?

**Comment by linsui** [comment]
No, it failed with the same problem that the days of week are not sorted. In fact, in our build they are sorted from SUNDAY to MONDAY but in upstream build they are not sorted.

**Comment by IzzySoft** [comment]
That is strange. It's sorted here fine (else it wouldn't be RB). As 1.8.6 was just released, maybe skip 1.8.5 and go straight to that? I doubt that 2 RBs in a row are coincidence, so I'd assume the issue fixed.

**Comment by pewaru-333** [comment]
I don't know the reason, because from version 1.8.4 I have setup the workflow, and the .apk is built from repo's code. Any ideas or suggestions maybe?

**Comment by IzzySoft** [comment]
I find it strange, too – especially as we cannot confirm the problem anymore. Maybe you can link to the recipe & log of your run, @linsui?

**Comment by IzzySoft** [comment]
That's really odd. You build with OpenJDK-17 on bullseye, same as we do, just with `assembleRelease` – from the very same commit. I see no explanation why our builds succeed RB without that sorting issue, while yours still have the sorting issue. I see you do [a bunch of additional gradle steps](https://gitlab.com/fdroid/fdroiddata/-/jobs/11267006138#L519) we don't, but that should not have this outcome.

Our recipe didn't change either, expect for adjusting the apksigner padding. So no clue why it fails for you…

**Comment by pewaru-333** [comment]
@linsui, hello. Same issue this time (1.8.7 version)?

**Comment by linsui** [comment]
Still the same issue happening randomly.

**Comment by IzzySoft** [comment]
1.8.7 was running through RB here fine, according to our logs.

> You build with OpenJDK-17 on bullseye, same as we do

Oops, correction: I of course meant bookworm there. bullseye we only use when JDK-11 is needed…

**Comment by pewaru-333** [comment]
Nice, the last version is on F-Droid repo. I will close this issue, but if the problem returns, please, reopen it.

**Comment by pewaru-333** [comment]
@linsui, work in progress...

**Comment by IzzySoft** [comment]
> Confirmed that it's reproducible sometimes.

So still lottery, and I've just drawn a "lucky number" on the first try? Curious for the next releases then…

**Comment by IzzySoft** [comment]
OK, so still flaky. Well, we'll take a look at the diff should that happen here again. Thanks!
